glusterd2/transaction/transaction.go

```diff
@@ -39,19 +42,43 @@ type Txn struct {
 // NewTxn returns an initialized Txn without any steps
 func NewTxn(ctx context.Context) *Txn {
 	t := new(Txn)
+
 	t.id = uuid.NewRandom()
 	t.reqID = gdctx.GetReqID(ctx)
+	t.locks = make(map[string]*concurrency.Mutex)
+
 	prefix := txnPrefix + t.id.String()
 	t.Ctx = NewCtxWithLogFields(log.Fields{
 		"txnid": t.id.String(),
 		"reqid": t.reqID.String(),
 	}).WithPrefix(prefix)
 
+	t.Ctx.Logger().Debug("new transaction created")
+	expTxn.Add("initiated_txn_in_progress", 1)
 	return t
 }
 
-// Cleanup cleans the leftovers after a transaction ends
-func (t *Txn) Cleanup() {
+// NewTxnWithLocks returns an empty Txn with locks obtained on given lockIDs
+func NewTxnWithLocks(ctx context.Context, lockIDs ...string) (*Txn, error) {
+	t := NewTxn(ctx)
+
+	for _, id := range lockIDs {
+		if err := t.Lock(id); err != nil {
+			t.Done()
+			return nil, err
+		}
+	}
+
+	return t, nil
+}
+
+// Done releases any obtained locks and cleans up the transaction namespace
+// Done must be called after a transaction ends
+func (t *Txn) Done() {
+	// Release obtained locks
+	for _, locker := range t.locks {
```

Comment:
Does the ordering in list of locks matter ? I was wondering whether the last lock taken should be the first to get unlocked.

Reply:
Not really ATM, and I don't expect it to. The locks are taken for the lifetime of the period of the transaction, and are released all at once at the end. I don't expect it to change.